src/test/java/org/ethereum/beacon/discovery/TestUtil.java

```diff
@@ -125,4 +129,46 @@ public static boolean isFieldsEqual(
         && that.getIdSignature().equals(other.getIdSignature())
         && that.getNodeRecord(nodeRecordFactory).equals(other.getNodeRecord(nodeRecordFactory));
   }
+
+  public static void waitFor(final CompletableFuture<?> future) throws Exception {
+    waitFor(future, 30);
+  }
+
+  public static void waitFor(final CompletableFuture<?> future, final int timeout)
+      throws Exception {
+    future.get(timeout, TimeUnit.SECONDS);
+  }
+
+  public static void waitFor(final ThrowingRunnable assertion) throws Exception {
+    int attempts = 0;
+    while (true) {
+      try {
+        assertion.run();
+        return;
+      } catch (Throwable t) {
+        if (attempts < 60) {
+          attempts++;
+          Thread.sleep(1000);
```

Comment:
for future it'd be nice to have a stub timer and not have to thread.sleep like this maybe, but its probably not worth doing now... just a thought.

Reply:
This is an actual waiter for tests that are actually using multiple threads so it really does need to actually wait.  I think it's actually only used by `DiscoveryIntegrationTest` in the end and used to be a private method there but got moved because it looked like it would be useful elsewhere. Didn't move it back because it is a handy method to have around.